### ISSUE TITLE
Connection reset by server while client is asleep

### ISSUE THREAD

**Issue by TomiBelan** [tailscale/tailscale#5021]
### What is the issue?

I use ssh to connect from my laptop to a server over Tailscale. When I suspend the laptop, the connection is reset after some time. When I resume the laptop and type any character, the client notices that the connection is broken and exits.

### Steps to reproduce

The client is a Windows laptop. I use various ssh clients (official ssh.exe, cygwin ssh.exe, libssh). I use `-o tcpkeepalive=no` ([manual](https://man.openbsd.org/ssh_config#TCPKeepAlive), [SE](https://unix.stackexchange.com/questions/34004/how-does-tcp-keepalive-work-in-ssh), [TLDP](https://tldp.org/HOWTO/TCP-Keepalive-HOWTO/usingkeepalive.html), [source](https://github.com/openssh/openssh-portable/blob/669a56bcfe73f8b985f2bba476ba834d55253acf/sshconnect.c#L525))

The server is a Linux machine running sshd from OpenSSH_7.6p1. The sshd_config is default. I am not using "Tailscale SSH". The server runs tailscaled unprivileged with `--tun=userspace-networking`, so sshd thinks it's a local TCP connection from 127.0.0.1.

1. Run the ssh client with the server's Tailscale IP and log in.
2. Put the laptop to sleep. Resume it after at least X hours (not sure about exact threshold).
3. Type any character in the ssh client. It prints: `client_loop: send disconnect: Connection reset by peer` and exits.

If I connect directly (not to Tailscale IP) immediately after resuming the laptop, and run `ps x`, it reveals that the `sshd: user@pts/N` process and the shell are already dead. I think this means that at some point while the laptop was asleep, either sshd or tailscaled tried to ping the client, timed out, and closed the connection. When the client sends a next packet, it receives RST.

I expect this will be hard to reproduce and you'll need more help from me to debug this. Is there anything else I should try or any information I can give you?

### Are there any recent changes that introduced the issue?

It used to work - the ssh connection used to survive overnight, as long as the shell was idle. The most likely recent change is that I upgraded the server from Tailscale 1.16.0 to 1.26.0.

When I downgraded the server side back to 1.16.0, it works again. That could be a coincidence, but it strongly suggests this is a regression between 1.16.0 and 1.26.0.

I'll try bisecting the version if I have time. But it will take a while. It doesn't happen with short sleeps, so I only get one try per day.

### OS

Linux, Windows

### OS version

Client: Windows 10. Server: Linux Mint 19.3

### Tailscale version

Client: 1.26.0. Server: 1.26.0 (works in 1.16.0)

### Bug report

BUG-b83d3ce1799672228097e3bda4a23d0cd15f9e7efe7290cd84e8c62563f56cee-20220708150118Z-8b7ce0a18c425364

**Comment by DentonGentry** [comment]
With 1.26.x are you using `tailscale up --ssh`?

With 1.16 the SSH server is external, most likely OpenSSHd. With 1.26.x there is the option to use https://tailscale.com/blog/tailscale-ssh/, and may lead to different behaviors than OpenSSHd.

**Comment by TomiBelan** [comment]
No, I am not using `--ssh`, with either version. In both cases I use the sshd from OpenSSH_7.6p1.

**Comment by DentonGentry** [comment]
This could be a side-effect of the fix for https://github.com/tailscale/tailscale/issues/4522, which went into 1.24.1. That would imply that 1.24.0 would not lose the SSH connection after a long idle time, while 1.24.1 would.

**Comment by bradfitz** [comment]
The issue is probably the userspace networking packet copy which has a timeout. But I don't think that's changed recently. We can go look.

**Comment by bradfitz** [comment]
Whoops, didn't see Denton's reply come in meanwhile. I forgot about that!

**Comment by TomiBelan** [comment]
Bisecting server versions progress report: 1.20.4 and 1.22.2 works, 1.24.1 breaks. I'll try 1.24.0 next. But this is already promising evidence for Denton's idea.

I also tried a ncat-based echo server, attempting to prove the issue is not related to ssh.
On the server, inside a tmux: `ncat -l 8400 -k -c cat`
On the client: `tailscale nc servername 8400`
But I didn't get the result I expected! With tailscaled=1.24.1 the ssh connection broke but the ncat connection stayed connected, and the cat process stayed running. How confusing! What makes them different? Next time I'll try sleeping the client for a longer time, in case it was just bad luck.

**Comment by TomiBelan** [comment]
As you suspected, 1.24.0 works and 1.24.1 fails. Thanks Denton!!

This time, I tried three connections - `ssh`, `tailscale nc servername 8400`, and `telnet servername 8400` (just in case `tailscale nc` is different). All three broke with tailscaled 1.24.1. That proves the problem is not specific to the ssh protocol or sshd.

I checked the timing using cron (`* * * * * (date; ps x) >> /tmp/ps`) to see when the relevant processes died. The connections broke after different durations. The sshd+bash and one ncat+cat died 131 minutes since connecting. The other ncat+cat died after 203 minutes. That's a bit surprising because I expected it to also stop after ~2 hours (the default TCP keepalive time). Anyway, I hope it helps tell you how long to keep your client (laptop/VM) suspended to reproduce it.

I wonder if there's any solution that fixes both #4522 and this issue, but I don't have any better ideas than letting the user manually configure it (ew). Re #4522 """I don't know if there is a way for userspace mode to detect peer keepalive tuning - that's something we can look into.""" - That probably wouldn't help me anyway, because OpenSSH's sshd usually enables keepalive with its default sshd_config, and I was relying on the happy coincidence that this keepalive used to only apply to the final local TCP hop between tailscaled and sshd. 

For now I'll just downgrade, it's not a high priority bug for me.
